go/src/koding/kites/kloud/provider/aws/stack.go

```diff
@@ -1,218 +1,268 @@
-package awsprovider
+package aws
 
 import (
-	"errors"
+	"bytes"
 	"fmt"
-	"strings"
-
-	"github.com/aws/aws-sdk-go/aws"
-	"github.com/aws/aws-sdk-go/aws/credentials"
-	"github.com/aws/aws-sdk-go/aws/session"
-	"github.com/aws/aws-sdk-go/service/iam"
+	"strconv"
+	"text/template"
+	"time"
 
 	"koding/kites/kloud/api/amazon"
-	"koding/kites/kloud/provider"
 	"koding/kites/kloud/stack"
-	"koding/kites/kloud/stackplan"
-
-	"golang.org/x/net/context"
+	"koding/kites/kloud/stack/provider"
+	"koding/kites/kloud/userdata"
 )
 
-// Cred represents jCredentialDatas.meta for "aws" provider.
-type Cred struct {
-	Region    string `json:"region" bson:"region" hcl:"region"`
-	AccessKey string `json:"access_key" bson:"access_key" hcl:"access_key"`
-	SecretKey string `json:"secret_key" bson:"secret_key" hcl:"secret_key"`
-
-	// Bootstrap metadata.
-	ACL       string `json:"acl,omitempty" bson:"acl,omitempty" hcl:"acl"`
-	CidrBlock string `json:"cidr_block,omitempty" bson:"cidr_block,omitempty" hcl:"cidr_block"`
-	IGW       string `json:"igw,omitempty" bson:"igw,omitempty" hcl:"igw"`
-	KeyPair   string `json:"key_pair,omitempty" bson:"key_pair,omitempty" hcl:"key_pair"`
-	RTB       string `json:"rtb,omitempty" bson:"rtb,omitempty" hcl:"rtb"`
-	SG        string `json:"sg,omitempty" bson:"sg,omitempty" hcl:"sg"`
-	Subnet    string `json:"subnet,omitempty" bson:"subnet,omitempty" hcl:"subnet"`
-	VPC       string `json:"vpc,omitempty" bson:"vpc,omitempty" hcl:"vpc"`
-	AMI       string `json:"ami,omitempty" bson:"ami,omitempty" hcl:"ami"`
+//go:generate $GOPATH/bin/go-bindata -mode 420 -modtime 1470666525 -pkg aws -o bootstrap.json.tmpl.go bootstrap.json.tmpl
+//go:generate go fmt bootstrap.json.tmpl.go
+
+var bootstrap = template.Must(template.New("").Parse(mustAsset("bootstrap.json.tmpl")))
+
+type bootstrapConfig struct {
+	AvailabilityZone string
+	KeyPairName      string
+	PublicKey        string
+	EnvironmentName  string
+}
+
+// Stack implements the stackplan.Stack interface.
+type Stack struct {
+	*provider.BaseStack
 }
 
-var _ stack.Validator = (*Cred)(nil)
+var _ provider.Stack = (*Stack)(nil)
 
-func (meta *Cred) BootstrapValid() error {
-	if meta.ACL == "" {
-		return errors.New("acl is empty or missing")
+func (s *Stack) VerifyCredential(c *stack.Credential) error {
+	cred := c.Credential.(*Cred)
+
+	if err := cred.Valid(); err != nil {
+		return err
 	}
-	if meta.CidrBlock == "" {
-		return errors.New("CIDR block is empty or missing")
+
+	_, err := amazon.NewClient(cred.Options())
```

Comment:
There's a bigger story about that - for now we require AWS creds to have full EC2 support + IAM::GetUser permission (to read amazon account ID). That's how we document requirement for AWS credential. I wouldn't want only to make EC2 checks here, I'd want to drop this requirement, so it's possible to run AWS stacks with less privileged user. This however requires a number of patches. Created https://github.com/koding/koding/issues/9137 to track this.